### PR TITLE
Separate regular PhD output from error handling

### PR DESCRIPTION
**Pull request by haszi**
Move error handling code into a class and remove PhD message output handling from it.
Introduce a new class to handle PhD message output.
Make the implicit dependency on the output functionality of classes explicit.
Update PEAR package.xml.
Fix tests.

### CODE REVIEW

File: phpdotnet/phd/Reader.php [hunk 1]
```diff
@@ -24,7 +27,7 @@ public function readContent(?string $node = null): string {
             if (\LIBXML_VERSION >= 20620) {
                 return self::readString();
             }
-            v("You are using libxml2 v%d, but v20620 or newer is preferred", \LIBXML_VERSION, VERBOSE_OLD_LIBXML);
+            $this->outputHandler->v("You are using libxml2 v%d, but v20620 or newer is preferred", \LIBXML_VERSION, VERBOSE_OLD_LIBXML);
 
             $node = $this->name;
```
Comment (Girgias) [inline phpdotnet/phd/Reader.php:32]: Aside, how old/recent is this version? Might make sense to drop this in a follow-up.

File: phpdotnet/phd/OutputHandler.php [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        $timestamp = $this->term_color(\sprintf("[%s - %s]", $time, $infoOrErrorString), $color);
+
+        return \fprintf($outputStream, "%s %s\n", $timestamp, $msg);
+    }
+    
+    /**
+     * Print info messages: v("printf-format-text" [, $arg1, ...], $verbose-level)
+     */
+    public function v(): bool {
+        $args = \func_get_args();
```
Comment (Girgias) [inline phpdotnet/phd/OutputHandler.php:64]: Use proper variadic arguments?
```suggestion
    public function v(...$args): bool {
```

File: phpdotnet/phd/OutputHandler.php [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+            VERBOSE_TOC_WRITING           => 'Writing TOC           ',
+            VERBOSE_CHUNK_WRITING         => 'Writing Chunk         ',
+            VERBOSE_MESSAGES              => 'Heads up              ',
+
+            // PhD warnings
+            VERBOSE_NOVERSION             => 'No version information',
+            VERBOSE_BROKEN_LINKS          => 'Broken links          ',
+            VERBOSE_OLD_LIBXML            => 'Old libxml2           ',
+            VERBOSE_MISSING_ATTRIBUTES    => 'Missing attributes    ',
+        ];
```
Comment (Girgias) [inline phpdotnet/phd/OutputHandler.php:85]: Should this be a class constant?

File: phpdotnet/phd/OutputHandler.php [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    public function __construct(
+        private Config $config
+    ) {}
+
+    /**
+     * Method to get a color escape sequence
+     */
+    private function term_color(string $text, string|bool $color): string {
+        return $this->config->color_output() && $color !== false ? "\033[" . $color . "m" . $text . "\033[m" : $text;
+    }
```
Comment (Girgias) [inline phpdotnet/phd/OutputHandler.php:15]: Is this `string|bool` actually `string|false` ?

File: phpdotnet/phd/OutputHandler.php [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+    public function printPhdInfo(string $msg, ?string $info = ""): int {
+        $color = $this->config->phd_info_color();
+        $outputStream = $this->config->phd_info_output();
+
+        return $this->print($msg, $outputStream, $color, $info);
+    }
```
Comment (Girgias) [inline phpdotnet/phd/OutputHandler.php:22]: Does `$info` need to be nullable if it has a default value of empty string?

File: phpdotnet/phd/OutputHandler.php [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+    private function printPhdWarning(string $msg, ?string $warning = ""): int {
```
Comment (Girgias) [inline phpdotnet/phd/OutputHandler.php:24]: Ditto `$warning`

File: phpdotnet/phd/OutputHandler.php [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        $color = $this->config->phd_warning_color();
+        $outputStream = $this->config->phd_warning_output();
+
+        return $this->print($msg, $outputStream, $color, $warning);
+    }
+
+    public function printUserError(string $msg, string $file, int $line, ?string $error = ""): int {
```
Comment (Girgias) [inline phpdotnet/phd/OutputHandler.php:31]: Ditto

File: phpdotnet/phd/OutputHandler.php [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+        $color = $this->config->user_error_color();
+        $outputStream = $this->config->user_error_output();
+        $data = \sprintf("%s:%d\n\t%s", $file, $line, $msg);
+
+        return $this->print($data, $outputStream, $color, $error);
+    }
+
+    public function printPhpError(string $msg, string $file, int $line, ?string $error = ""): int {
```
Comment (Girgias) [inline phpdotnet/phd/OutputHandler.php:39]: Ditto

File: phpdotnet/phd/OutputHandler.php [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+        $color = $this->config->php_error_color();
+        $outputStream = $this->config->php_error_output();
+        $data = \sprintf("%s:%d\n\t%s", $file, $line, $msg);
+
+        return $this->print($data, $outputStream, $color, $error);
+    }
+
+    private function print(string $msg, $outputStream, string|false $color = false, ?string $infoOrErrorString = ""): int {
```
Comment (Girgias) [inline phpdotnet/phd/OutputHandler.php:47]: Ditto for `$infoOrErrorString`, especially as it doesn't seem like you are handling the `null` case at all.

File: phpdotnet/phd/ErrorHandler.php [hunk 10]
```diff
@@ -0,0 +1,71 @@
+<?php
+namespace phpdotnet\phd;
+
+class ErrorHandler
+{
+    public function __construct(
+        private OutputHandler $outputHandler
+    ) {}
+
+    public function errh($errno, $msg, $file, $line) {
```
Comment (Girgias) [inline phpdotnet/phd/ErrorHandler.php:10]: :| please name this something better

File: phpdotnet/phd/ErrorHandler.php [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+            E_STRICT                      => 'E_STRICT              ',
+            E_WARNING                     => 'E_WARNING             ',
+            E_NOTICE                      => 'E_NOTICE              ',
+
+            // User Triggered Errors
+            E_USER_ERROR                  => 'E_USER_ERROR          ',
+            E_USER_WARNING                => 'E_USER_WARNING        ',
+            E_USER_NOTICE                 => 'E_USER_NOTICE         ',
+            E_USER_DEPRECATED             => 'E_USER_DEPRECATED     ',
+        );
```
Comment (Girgias) [inline phpdotnet/phd/ErrorHandler.php:24]: Class constant?

File: phpdotnet/phd/ErrorHandler.php [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+        static $recursive = false;
```
Comment (Girgias) [inline phpdotnet/phd/ErrorHandler.php:25]: This could be a property now, but that is a minor nit.

File: phpdotnet/phd/Reader.php (hunk 1 above)
Comment (haszi) [reply phpdotnet/phd/Reader.php:32]: It's from July of 2005. :-)

File: tests/setup.php [hunk 13]
```diff
@@ -3,13 +3,19 @@
...
-require_once __INSTALLDIR__ . DIRECTORY_SEPARATOR . "phpdotnet" . DIRECTORY_SEPARATOR . "phd" . DIRECTORY_SEPARATOR . "functions.php";
-
 $config = new Config;
 
+$outputHandler = new OutputHandler($config);
+
+$errorHandler = new ErrorHandler($outputHandler);
+$olderrrep = error_reporting();
+error_reporting($olderrrep | VERBOSE_DEFAULT);
+set_error_handler([$errorHandler, "handleError"]);
```
Comment (Girgias) [inline tests/setup.php:17]: Do we require PHP 8.1 or not? As could use:
```suggestion
set_error_handler($errorHandler::handleError(...));
```
if yes

File: render.php [hunk 14]
```diff
@@ -4,18 +4,24 @@
...
 spl_autoload_register(array(__NAMESPACE__ . "\\Autoloader", "autoload"));
 
 $config = new Config;
 
+$outputHandler = new OutputHandler($config);
+
+$errorHandler = new ErrorHandler($outputHandler);
+$olderrrep = error_reporting();
+error_reporting($olderrrep | VERBOSE_DEFAULT);
+set_error_handler([$errorHandler, "handleError"]);
```
Comment (Girgias) [inline render.php:18]: Ditto about the first class callable syntax.

**Review by Girgias — APPROVED** [review]
Minor final comment, but LGTM thank you!

**Comment by haszi** [comment]
Test failure is PHP 8.0 tests failing due the first-class callable syntax. We already had `never` return types so I'm really curious how they were not failing before. I'm opening a separate PR to remove 8.0 from the test matrix.

**Pull request merged**